pkg/prometheus/operator.go

```diff
@@ -1362,6 +1452,164 @@ func (c *Operator) sync(ctx context.Context, key string) error {
+		pStatus.UpdatedReplicas += int32(len(stsReporter.Updated()))
+		pStatus.AvailableReplicas += int32(len(stsReporter.Ready()))
+		pStatus.UnavailableReplicas += int32(len(stsReporter.pods) - len(stsReporter.Ready()))
+
+		pStatus.ShardStatuses = append(
+			pStatus.ShardStatuses,
+			monitoringv1.ShardStatus{
+				ShardID:             strconv.Itoa(shard),
+				Replicas:            int32(len(stsReporter.pods)),
+				UpdatedReplicas:     int32(len(stsReporter.Updated())),
+				AvailableReplicas:   int32(len(stsReporter.Ready())),
+				UnavailableReplicas: int32(len(stsReporter.pods) - len(stsReporter.Ready())),
+			},
+		)
+
+		if len(stsReporter.Ready()) == len(stsReporter.pods) {
+			// All pods are ready (or the desired number of replicas is zero).
+			continue
+		}
+
+		if len(stsReporter.Ready()) == 0 {
+			availableCondition.Reason = "NoPodReady"
+			availableCondition.Status = monitoringv1.PrometheusConditionFalse
+		} else if availableCondition.Status != monitoringv1.PrometheusConditionFalse {
+			availableCondition.Reason = "SomePodsNotReady"
+			availableCondition.Status = monitoringv1.PrometheusConditionDegraded
+		}
+
+		for _, p := range stsReporter.pods {
+			if m := p.Message(); m != "" {
+				messages = append(messages, fmt.Sprintf("shard %d: pod %s: %s", shard, p.Name, m))
+			}
+		}
+	}
+
+	availableCondition.Message = strings.Join(messages, "\n")
+
+	// Compute the Reconciled ConditionType.
+	reconciledCondition := monitoringv1.PrometheusCondition{
+		Type:   monitoringv1.PrometheusReconciled,
+		Status: monitoringv1.PrometheusConditionTrue,
+		LastTransitionTime: metav1.Time{
+			Time: time.Now().UTC(),
+		},
+	}
+	reconciliationStatus, found := c.reconciliations.GetStatus(key)
+	if !found {
+		reconciledCondition.Status = monitoringv1.PrometheusConditionUnknown
+	} else if !reconciliationStatus.Ok() {
+		reconciledCondition.Status = monitoringv1.PrometheusConditionFalse
+		reconciledCondition.Reason = reconciliationStatus.Reason()
+		reconciledCondition.Message = reconciliationStatus.Message()
+	}
+
+	// Update the last transition times only if the status of the available condition has changed.
+	for _, condition := range p.Status.Conditions {
+		if condition.Type == availableCondition.Type && condition.Status == availableCondition.Status {
+			availableCondition.LastTransitionTime = condition.LastTransitionTime
+			continue
+		}
+
+		if condition.Type == reconciledCondition.Type && condition.Status == reconciledCondition.Status {
+			reconciledCondition.LastTransitionTime = condition.LastTransitionTime
+		}
+	}
+
+	pStatus.Conditions = append(pStatus.Conditions, availableCondition, reconciledCondition)
+
+	b, err := json.Marshal(&monitoringv1.Prometheus{
+		TypeMeta: p.TypeMeta,
+		ObjectMeta: metav1.ObjectMeta{
+			Name: p.Name,
+		},
+		Status: pStatus,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal object")
+	}
+
+	_, err = c.mclient.MonitoringV1().Prometheuses(p.Namespace).Patch(
```

Comment:
I think it is safer to use a regular `Update` instead of a server-side apply. The operator should be the only manager of the CR status and I don't think we should have other managers for any status fields.